integrations/pull_merge_test.go

```diff
@@ -61,31 +61,75 @@ func testPullCleanUp(t *testing.T, session *TestSession, user, repo, pullnum str
 
 func TestPullMerge(t *testing.T) {
 	onGiteaRun(t, func(t *testing.T, giteaURL *url.URL) {
-		hookTasks, err := models.HookTasks(1, 1) //Retrieve previous hook number
-		assert.NoError(t, err)
-		hookTasksLenBefore := len(hookTasks)
+		createPullNotified, deferableCreate := notifierListener.RegisterChannel("NotifyNewPullRequest", 0, &models.PullRequest{})
+		defer deferableCreate()
+
+		mergePullNotified, deferableMerge := notifierListener.RegisterChannel("NotifyMergePullRequest", 0, &models.PullRequest{})
+		defer deferableMerge()
 
 		session := loginUser(t, "user1")
 		testRepoFork(t, session, "user2", "repo1", "user1", "repo1")
 		testEditFile(t, session, "user1", "repo1", "master", "README.md", "Hello, World (Edited)\n")
 
+		var prInterface interface{}
+
 		resp := testPullCreate(t, session, "user1", "repo1", "master", "This is a pull title")
+		select {
+		case prInterface = <-createPullNotified:
+		case <-time.After(500 * time.Millisecond):
+			assert.Fail(t, "Took too long to notify!")
+		}
+		pr := prInterface.(*models.PullRequest)
+		pr.LoadBaseRepo()
+		pr.LoadHeadRepo()
+		pr.BaseRepo.MustOwner()
+		pr.HeadRepo.MustOwner()
+
+		assert.EqualValues(t, "user1", pr.HeadRepo.Owner.Name)
+		assert.EqualValues(t, "repo1", pr.HeadRepo.Name)
+		assert.EqualValues(t, "user2", pr.BaseRepo.Owner.Name)
+		assert.EqualValues(t, "repo1", pr.BaseRepo.Name)
 
 		elem := strings.Split(test.RedirectURL(resp), "/")
 		assert.EqualValues(t, "pulls", elem[3])
+
 		testPullMerge(t, session, elem[1], elem[2], elem[4], models.MergeStyleMerge)
 
-		hookTasks, err = models.HookTasks(1, 1)
-		assert.NoError(t, err)
-		assert.Len(t, hookTasks, hookTasksLenBefore+1)
+		select {
+		case prInterface = <-mergePullNotified:
+		case <-time.After(500 * time.Millisecond):
+			assert.Fail(t, "Took too long to notify!")
+		}
+
+		pr = prInterface.(*models.PullRequest)
+		pr.LoadBaseRepo()
+		pr.LoadHeadRepo()
+		pr.BaseRepo.MustOwner()
+		pr.HeadRepo.MustOwner()
+
+		assert.EqualValues(t, "user1", pr.HeadRepo.Owner.Name)
+		assert.EqualValues(t, "repo1", pr.HeadRepo.Name)
+		assert.EqualValues(t, "user2", pr.BaseRepo.Owner.Name)
+		assert.EqualValues(t, "repo1", pr.BaseRepo.Name)
+
+		time.Sleep(100 * time.Millisecond)
```

Comment:
Why is this `Sleep` required?

Reply:
I've changed this whole section of the test to be a bit clearer.